src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -2247,11 +2255,67 @@ internal void OnConnectionClosed()
             }
         }
 
-        private void OnTimeout(object state)
+        public void SetTimeoutStateStopped()
+        {
+            Interlocked.Exchange(ref _timeoutState, TimeoutStopped);
+            _timeoutIdentityValue = 0;
+        }
+
+        public bool IsTimeoutStateExpired
         {
-            if (!_internalTimeout)
+            get
             {
-                _internalTimeout = true;
+                int state = _timeoutState;
+                return state == TimeoutExpiredAsync || state == TimeoutExpiredSync;
+            }
+        }
+
+        private sealed class TimeoutState
+        {
+            public int IdentityValue { get; set; }
+        }
+
+        private void OnTimeoutAsync(object state)
+        {
+#if DEBUG
+            Thread.Sleep(13000);
```

Comment:
Note: remember to remove this if we intend to merge this PR

Reply:
I've address this in my PR over this one.